File: baekjoon-online-judge/platinum/01214.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws Exception  {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		// Input
		String[] in = br.readLine().split(" ");
		long D = Long.parseLong(in[0]);
		long P = Long.parseLong(in[1]);
		long Q = Long.parseLong(in[2]);

		// Solve
		long max = Math.max(P, Q);
		long min = Math.min(P, Q);
		P = max; Q = min;

		long k = Math.min((int) 1e9, (P - (D % P)) % P);
		long N = Math.min(D / P, Q);
		for (int x = 0; x <= N; ++x) {
			k = Math.min(k, (Q - (D - P*x) % Q) % Q);
		}

		// Output
		System.out.print(D + k);
	}
}

File: jungol/Intermediate/01214.py
from sys import stdin

input = stdin.readline

if __name__ == '__main__':

    # Input
    _, *H = map(int, input().split())

    # Solve
    answer = 0
    H.append(0)
    stack = [] # idx, H[idx]

    for idx, height in enumerate(H):
        start_idx = idx

        while stack and stack[-1][1] > height: # prev_height > height
            prev_idx, prev_height = stack.pop()
            answer = max(answer, prev_height * (idx - prev_idx))
            start_idx = prev_idx
        stack.append((start_idx, height))

    # Output
    print(answer)